nidirect_custom_blocks/src/Plugin/Block/NIDirectArticleTeasersByTopic.php

```diff
@@ -0,0 +1,157 @@
+use Drupal\Core\Url;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'NIDirectArticleTeasersByTopic' block.
+ *
+ * @Block(
+ *  id = "nidirect_article_teasers_by_topic",
+ *  admin_label = @Translation("NIDirect Article Teasers by Topic"),
+ * )
+ */
+class NIDirectArticleTeasersByTopic extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The renderer.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * The route match service.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * Constructs an AggregatorFeedBlock object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   Current route match service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, RendererInterface $renderer = NULL, RouteMatchInterface $route_match) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->renderer = $renderer;
+    $this->routeMatch = $route_match;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('renderer'),
+      $container->get('current_route_match')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    $build = [];
+    $cache_tags = [];
+    // Get the current node.
+    $node = $this->routeMatch->getParameter('node');
+    // Add custom cache tag for taxonomy term listing.
+    $cache_tags[] = 'taxonomy_term_list:' . $node->field_subtheme->target_id;
+    // Get a list of article teasers by term.
+    $results = $this->renderArticleTeasersByTerm($node->field_subtheme->target_id, $node->id(), $cache_tags);
+    // Get a list of article teasers by topic.
+    $results += $this->renderArticleTeasersByTopic($node->field_subtheme->target_id, $cache_tags);
+    // Sort entries alphabetically (regardless of type).
+    ksort($results);
+    // Will be processed by block--nidirect-article-teasers-by-topic.html.twig.
+    $build['nidirect_article_teasers_by_topic'] = $results;
+    $build['nidirect_article_teasers_by_topic']['#cache'] = [
+      'tags' => $cache_tags,
+    ];
+    return $build;
+  }
+
+  /**
+   * Utility function to render 'articles by term' view.
+   */
+  private function renderArticleTeasersByTerm($tid, $current_nid, &$cache_tags) {
```

Comment:
Not critical, but if you put type hints before each parameter it helps autodoc tools and PHPStorm to detect any inconsistencies before runtime. I'd suggest something like `(int $tid, int $current_nid, array &$cache_tags)`

Reply:
Arg! That's one that you always spot - and one that I always forget !
Good point, done now